### ISSUE TITLE
Overfit training produces no detection

### ISSUE THREAD

**Issue by rlan** [thtrieu/darkflow#411]
I have been training with my own dataset from scratch without success (no bounding boxes). I read about #80 and #142. Taking @thtrieu 's advice, I went to the basics and tried a overfit, blank training using VOC images, but does not get any bounding boxes. As you know, this dataset already has working weights. Any help would be appreciated. Other people may have the same problem still, e.g. #409

Here are my steps:
Selected 3 images from Pascal VOC 2007: 1 image has person, 1 image has dog and 1 image as both.

- 000037.jpg
- 000546.jpg
- 001372.jpg

Step 1. To make sure images are detected by the trained network, I ran the given cfg/tiny-yolo-voc.cfg and @thtrieu 's tiny-yolo-voc.weights. The 3 images are indeed detected correct (with bounding boxes).

labels.txt
> person
> dog

Step 2. Then I create a new cfg to train for these two classes. I copied tiny-yolo-voc.cfg and modified:

> last convolutional
> filters=35
> [region]
> classes=2

I turned off the randomness in darkflow/net/yolo/predict.py:
> allobj = None; print("Randomness REMOVED") 
> if allobj is not None: # in training mode

I added more tensorflow summaries to monitor progress:

darkflow/net/yolov2/train.py
>     tf.summary.scalar('{} max adjusted_prob'.format(m['model']), tf.reduce_max(adjusted_prob))
>     tf.summary.scalar('{} max intersect'.format(m['model']), tf.reduce_max(intersect))
>     tf.summary.scalar('{} max best_box'.format(m['model']), tf.reduce_max(best_box))

Training command:

> ./flow --model cfg/new2-tiny-yolo-voc.cfg --train --lr 2e-3 --dataset ~/Pictures/voc_overfit/ --annotation ~/Pictures/voc_overfit/ --gpu 0.5 --trainer adam --verbalise --save 30 | tee log_new2_voc.txt

I stopped training at step 110:

> step 110 - loss 0.04657846689224243 - moving ave loss 0.08889748694728852

I check graphs on Tensorboard:
- adjusted_prob starts from 0.5 and goes to 1.0 around step 30.
- loss starts from 100 and goes to 5 around step 30.
- best box stays flat being t end at 1.
- max intersect flat begin to end at 88.

The two flat lines is suspicious. Any ideas to dig further?

Environment. MacOS Sierra 10.12.
Tensorflow, cython, numpy, opencv installed in a virtual environment. I can run the Tensorflow MNIST tutorial without problems.

**Comment by rlan** [comment]
@stanifrolov I am repeating VOC training from scratch using 3 images. I verified labels are correct (as you suggested in #393). What can I capture to show the network is converging?

**Comment by stanifrolov** [comment]
You could try flowing with --json and check whether or not you get any objects and check the confidence.

**Comment by andresmafladelgado** [comment]
I am having issues when training with my own data as well. Have you tried changing the threshold value on test images?

**Comment by rlan** [comment]
@stanifrolov @andresmafladelgado I took a big step back to darknet (https://github.com/pjreddie/darknet), which is the source of this port. I am able to train successfully there with new objects. When I get a chance, I will try your suggestions.

**Comment by SujeethJinesh** [comment]
Hi @rlan could you detail the steps you took to do that? I'm facing the same issue in this thread when training on the Udacity Self Driving car dataset. Thanks!

**Comment by rlan** [comment]
@SujeethJinesh Are you asking about darknet (not darkflow)? For dry run, I first reproduce Redmon's paper result using VOC dataset. Just follow his instruction on his website. If you use his pretrained imagenet weights as starting point, training will be faster. Then I label my own dataset to look like a VOC dataset and repeat the dry run steps. I get detections right after.

**Comment by junxuezheng** [comment]
@rlan i want to get the AP and mAP using VOC 2007,could you sent the code to my gmail(the matlab code that you modify).
my gmail junxuezheng@gmai.com
thanks.

**Comment by rlan** [comment]
@junxuezheng I hacked up my code and were not version controled. However, the core changes are in example_detector.m, before:
``
    detector=train(VOCopts,cls);                            % train detector
    test(VOCopts,cls,detector);                             % test detector
    [recall,prec,ap]=VOCevaldet(VOCopts,'comp3',cls,true);  % compute and display PR
``

after:
``
    %display(cls)
    %detector=train(VOCopts,cls);                            % train detector
    %test(VOCopts,cls,detector);                             % test detector
    [recall,prec,ap]=VOCevaldet(VOCopts,'comp4',cls,true);  % compute and display PR
``

Note comp3 vs comp4.

Cheers,

**Comment by junxuezheng** [comment]
hello Rick,
I am very glad to receive your reply.

I have do what you say,but i have a problem about function  VOCevaldet() .

[recall,prec,ap]=VOCevaldet(VOCopts,'comp4',cls,true); 


in
function [rec,prec,ap] = VOCevaldet(VOCopts,id,cls,draw)
% load results
[ids,confidence,b1,b2,b3,b4]=textread(sprintf(VOCopts.detrespath,id,cls),'%s %f %f %f %f %f');

1,What is "comp3" or "comp4" mean ?  we can see   "comp4 is "id " ,
so   
[ids,confidence,b1,b2,b3,b4]=textread('.....VOCdevkit/results/VOC2007/Main/comp4_det_test_cls.txt' ,%s %f %f %f %f %f');

2.As you know , darflow NET out put is one jason for every picture.
how can we get  "comp4_det_test_cls.txt "  from so many jason files?

**Comment by junxuezheng** [comment]
@rlan 
hello Rick,
I am very glad to receive your reply.

I have do what you say,but i have a problem about function  VOCevaldet() .

[recall,prec,ap]=VOCevaldet(VOCopts,'comp4',cls,true); 


in
function [rec,prec,ap] = VOCevaldet(VOCopts,id,cls,draw)
% load results
[ids,confidence,b1,b2,b3,b4]=textread(sprintf(VOCopts.detrespath,id,cls),'%s %f %f %f %f %f');

1,What is "comp3" or "comp4" mean ?  we can see   "comp4 is "id " ,
so   
[ids,confidence,b1,b2,b3,b4]=textread('.....VOCdevkit/results/VOC2007/Main/comp4_det_test_cls.txt' ,%s %f %f %f %f %f');

2.As you know , darflow NET out put is one jason for every picture.
how can we get  "comp4_det_test_cls.txt "  from so many jason files?

**Comment by rlan** [comment]
@junxuezheng Per my previous comments in this thread, I was **not** able to train darkflow with VOC. I was able to train darknet. darknet outputs "comp4_*.txt" files which the VOC Matlab file directly reads.

2) darkflow, instead, outputs json's. You may convert darkflow's json file to such comp4_*.txt. I did not look inside them. I am not sure if there are multiple detections per json. But darknet outputs all detections with non zero probabilities. All detections for all input images is in one file per class. Unlike json, the bounding box coordinates the Matlab file expects is normalized to image height and width. In the textread line, ids is file name. You may read the matlab code and figure out which corners are b1 b2 b3 and b4.

1) "comp4" vs "comp3" does not change any logic. Just a file name prefix.

Hope this helps,

**Comment by junxuezheng** [comment]
@rlan 
hey,
I am very glad to receive your reply.

Ok,I will convert darkflow's json file to  comp4_*.txt,
1. Can you send me one comp4_*.txt? I want to see what does it like in comp4_*.txt .

**Comment by rlan** [comment]
Hi @junxuezheng ,
Here is *comp4_det_test_aeroplane.txt* after running VOC-2007-trained model: https://pastebin.com/sSjnTipq

P.S. Sorry, I mentioned previously that the box corners are normalized values. Looking at above txt file, they are not. Will edit my previous comment.

**Comment by junxuezheng** [comment]
@rlan 
hi,I am very happy to receive your reply,

I have converted darkflow's json file to comp4_*.txt,and get the ap with your help.
thanks.
Good luck to you.
